File: ServerSide/BallPassSchedulePattern/Players_QrcodeTypes/efghijkl2_ChecksumPlayerDir/ChecksumPlayer.py
import os, sys
sys.path.append("../..")
from Players_CommonPlayers.SuperPlayerDir.SuperPlayer import SuperPlayer

class ChecksumPlayer(SuperPlayer):
    def __init__(self):
        super().__init__()  # スーパークラスの初期化メソッドを呼び出す
        self.my_name = None  # 必ずNoneで初期化
        self.self.mode_charNumInfo_checksum_bitlist = None
        self.loop11101100and00010001pad_only_list = None

    def return_my_name(self):
        return "ChecksumPlayer"

    def calculate_checksum(self, data_list):
        """
        データリストからChecksumを計算する
        :param data_list: 数値のリスト
        :return: Checksum値
        """
        return sum(data_list) % 256

    def append_checksum(self, data_list, checksum):
        """
        データリストにChecksumを追加する
        :param data_list: 数値のリスト
        :param checksum: 計算されたChecksum
        :return: Checksum付きのデータリスト
        """
        return data_list + [checksum]

    def print_data(self, data_list, checksum, data_list_with_checksum):
        """
        データリスト、Checksum、Checksum付きデータリストを出力する
        """
        print("元のデータリスト:", data_list)
        print("Checksum:", checksum)
        print("Checksum付きデータリスト:", data_list_with_checksum)

    def main(self):
        """
        mainメソッドで全てのメソッドを実行
        """
        
        """ 入力 """
        woP = self.one_time_world_instance.qRCodeBitConversionPlayer
        self.data_str = woP.data_bits # データのみ
        self.mode_charNumInfo_bitlist = woP.mode_and_countinfo_bit + self.data_str
        to_decimal = self.one_time_world_instance.polynomialDivisionPlayer.bit_list_to_decimal_list # to10bitメソッド
        
        """ メイン """
        # 任意のデータリスト
        mode_charNumInfo_decimallist = to_decimal(self.self.mode_charNumInfo_bitlist)
        
        # Checksumの計算
        checksum = self.calculate_checksum(mode_charNumInfo_decimallist)
        
        # Checksumをデータに追加
        self.mode_charNumInfo_checksum_bitlist  = self.append_checksum(self.mode_charNumInfo_bitlist , checksum)
        
        """ 出力 """
        self.mode_charNumInfo_checksum_bitlist
        
        """ プレイヤー自身を更新 """
        self.one_time_world_instance.checksumPlayer = self
        
        return "Completed"

File: InstanceSide/BallPassSchedulePattern/Players_QRcodeTypes/efghijkl_ChecksumPlayerDir/ChecksumPlayer.cs
/// Created by Masarina(Twitter_@Masarina002)

using UdonSharp;
using UnityEngine;

public class ChecksumPlayer : SuperPlayer
{
    public RinaNumpy rinaNumpy;  // RinaNumpyクラスのインスタンス
    public QRCodeBitConversionPlayer qRCodeBitConversionPlayer;
    public PolynomialDivisionPlayer polynomialDivisionPlayer;

    public string[] modeCharNumInfoChecksumBitlist;
    public string dataStr;
    public string modeCharNumInfoBitlist;

    public override string ReturnMyName()
    {
        return "ChecksumPlayer";
    }

    public int CalculateChecksum(int[] dataList)
    {
        // データリストからChecksumを計算し、256で割った余りを返す
        int sum = 0;
        for (int i = 0; i < dataList.Length; i++)
        {
            sum += dataList[i];
        }
        return sum % 256;
    }

    public int[] AppendChecksum(int[] dataList, int checksum)
    {
        // データリストにChecksumを追加
        int[] dataListWithChecksum = new int[dataList.Length + 1];
        RinaNumpy.CopyIntArray(dataList, dataListWithChecksum, dataList.Length); // RinaNumpyで配列をコピー
        dataListWithChecksum[dataList.Length] = checksum;
        return dataListWithChecksum;
    }

    private string ArrayToString(int[] array)
    {
        // 配列を文字列に変換
        int[] tempArray = new int[array.Length]; // コピー用の一時配列
        RinaNumpy.CopyIntArray(array, tempArray, array.Length); // 配列をコピー
        string result = "";

        for (int i = 0; i < tempArray.Length; i++)
        {
            result += tempArray[i].ToString(); // 各要素を文字列化して結合
            if (i < tempArray.Length - 1)
            {
                result += ", "; // 区切り文字を追加
            }
        }

        return result;
    }

    public void PrintData(int[] dataList, int checksum, int[] dataListWithChecksum)
    {
        // データリスト、Checksum、Checksum付きデータリストをログ出力
        Debug.Log("元のデータリスト: " + ArrayToString(dataList));
        Debug.Log("Checksum: " + checksum);
        Debug.Log("Checksum付きデータリスト: " + ArrayToString(dataListWithChecksum));
    }

    public override string ExecuteMain()
    {
        // QRCodeBitConversionPlayerからビットデータを取得
        this.dataStr = qRCodeBitConversionPlayer.dataBits;
        this.modeCharNumInfoBitlist = qRCodeBitConversionPlayer.modeAndCountInfoBit + this.dataStr;

        // PolynomialDivisionPlayerからビット列を10進数リストに変換
        int[] modeCharNumInfoDecimalList = polynomialDivisionPlayer.BitListToDecimalList(this.modeCharNumInfoBitlist);

        // Checksumの計算
        int checksum = CalculateChecksum(modeCharNumInfoDecimalList);

        // Checksumをデータリストに追加
        int[] modeCharNumInfoChecksumBitlist = AppendChecksum(modeCharNumInfoDecimalList, checksum);
        this.modeCharNumInfoChecksumBitlist = new string[modeCharNumInfoChecksumBitlist.Length];

        // 配列の内容を文字列形式で保存
        for (int i = 0; i < modeCharNumInfoChecksumBitlist.Length; i++)
        {
            this.modeCharNumInfoChecksumBitlist[i] = modeCharNumInfoChecksumBitlist[i].ToString();
        }

        return "Completed";
    }
}
